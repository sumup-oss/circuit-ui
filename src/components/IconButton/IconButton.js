/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import { size, hideVisually } from 'polished';

const buttonStyles = ({ theme }) => css`
  label: icon-button;
  padding: 0;
  margin: 0;
  display: inline-block;
  background-color: transparent;
  border: none;
  cursor: pointer;
  overflow: initial;
  color: ${theme.colors.n700};

  &:hover {
    color: ${theme.colors.n900};
  }

  &:focus,
  &:active {
    outline: none;
  }

  > svg {
    ${size('100%')};
  }
`;

const Button = styled('button')(buttonStyles);

const labelStyles = () => css`
  ${hideVisually()};
`;

const Label = styled('span')(labelStyles);

/**
 * Accessible icon button.
 */
const IconButton = ({ children, label, ...props }) => {
  if (!children) {
    return null;
  }
  const iconChild = Children.only(children);
  const icon = cloneElement(iconChild, { role: 'presentation' });
  return (
    <Button type="button" title={label} {...props}>
      {icon}
      {label && <Label>{label}</Label>}
    </Button>
  );
};

IconButton.propTypes = {
  /**
   * The icon element. Must be a single child.
   */
  children: PropTypes.element.isRequired,
  /**
   * Visually hidden text label for screen readers. Crucial for accessibility.
   */
  label: PropTypes.string.isRequired
};

/**
 * @component
 */
export default IconButton;
