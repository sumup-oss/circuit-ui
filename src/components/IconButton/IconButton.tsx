/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Children, cloneElement, FC, ReactElement } from 'react';
import { css } from '@emotion/core';
import { hideVisually } from 'polished';
import { Theme } from '@sumup/design-tokens';

import styled from '../../styles/styled';
import { Button, ButtonProps } from '../Button/Button';

export interface IconButtonProps extends ButtonProps {
  /**
   * A single icon element.
   */
  children: ReactElement;
  /**
   * Short label to describe the function of the button. Displayed as title
   * on hover, and accessible to screen readers.
   */
  label: string;
}

const Label = styled('span')(hideVisually);

const sizeStyles = (size: IconButtonProps['size'] = 'mega') => (
  theme: Theme
) => {
  const sizeMap = {
    kilo: theme.spacings.bit,
    mega: theme.spacings.byte,
    giga: theme.spacings.kilo
  };

  return css({
    label: `button--${size}`,
    padding: sizeMap[size]
  });
};

/**
 * The IconButton component displays a button with a single icon
 * as its only child.
 */
export const IconButton: FC<IconButtonProps> = ({
  children,
  label,
  size,
  ...props
}) => {
  const child = Children.only(children);
  const icon = cloneElement(child, { role: 'presentation' });
  return (
    <Button title={label} css={sizeStyles(size)} {...props}>
      {icon}
      <Label>{label}</Label>
    </Button>
  );
};
