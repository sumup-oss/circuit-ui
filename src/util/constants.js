/**
 * Position
 */

export const TOP = 'top';
export const BOTTOM = 'bottom';
export const LEFT = 'left';
export const RIGHT = 'right';

/**
 * Alignment
 */

export const START = 'start';
export const END = 'end';
export const CENTER = 'center';

/**
 * Sizes
 */

export const BIT = 'bit';
export const BYTE = 'byte';
export const KILO = 'kilo';
export const MEGA = 'mega';
export const GIGA = 'giga';
export const TERA = 'tera';
export const PETA = 'peta';
export const EXA = 'exa';
export const ZETTA = 'zetta';

/**
 * Shadows
 */

export const SINGLE = 'single';
export const DOUBLE = 'double';
export const TRIPLE = 'triple';

/**
 * Colors?
 */

export const SUCCESS = 'success';
export const DANGER = 'danger';
export const WARNING = 'warning';
