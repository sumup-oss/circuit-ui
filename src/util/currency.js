/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { compose, defaultTo, get, isNumber, isString } from 'lodash/fp';

import { NUMBER_SEPARATORS, formatNumber, getNumberFormat } from './numbers';
import { currencyToRegex } from './regex';

export const CURRENCY_FORMATS = {
  EUR: {
    'el-CY': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'el-GR': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'et-EE': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'de-AT': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'de-DE': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'de-LU': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'en-IE': { prependSymbol: true, fractionalPrecision: 2, addSpace: false },
    'en-MT': { prependSymbol: true, fractionalPrecision: 2, addSpace: false },
    'es-ES': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'fi-FI': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'fr-BE': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'fr-FR': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'fr-LU': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'it-IT': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'lt-LT': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'lv-LV': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'nl-BE': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    'nl-NL': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    'pt-PT': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'sk-SK': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'sl-SI': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  CHF: {
    'de-CH': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    'it-CH': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    'fr-CH': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: true, fractionalPrecision: 2, addSpace: false }
  },
  GBP: {
    'en-GB': { prependSymbol: true, fractionalPrecision: 2, addSpace: false },
    default: { prependSymbol: true, fractionalPrecision: 2, addSpace: false }
  },
  PLN: {
    'pl-PL': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  RUB: {
    'ru-RU': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  BRL: {
    'pt-BR': { prependSymbol: true, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: true, fractionalPrecision: 2, addSpace: true }
  },
  SEK: {
    'sv-SE': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  USD: {
    'en-US': { prependSymbol: true, fractionalPrecision: 2, addSpace: false },
    default: { prependSymbol: true, fractionalPrecision: 2, addSpace: false }
  },
  CLP: {
    'es-CL': { prependSymbol: true, fractionalPrecision: 0, addSpace: true },
    default: { prependSymbol: true, fractionalPrecision: 0, addSpace: true }
  },
  BGN: {
    'bg-BG': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  CZK: {
    'cs-CZ': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  DKK: {
    'da-DK': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  HUF: {
    'hu-HU': { prependSymbol: false, fractionalPrecision: 0, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 0, addSpace: true }
  },
  NOK: {
    'nb-NO': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    'nn-NO': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  RON: {
    'ro-RO': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  HRK: {
    'hr-HR': { prependSymbol: false, fractionalPrecision: 2, addSpace: true },
    default: { prependSymbol: false, fractionalPrecision: 2, addSpace: true }
  },
  COP: {
    default: { prependSymbol: true, fractionalPrecision: 2, addSpace: false },
    'es-CO': { prependSymbol: true, fractionalPrecision: 2, addSpace: false }
  }
};

export const CURRENCY_SYMBOLS = {
  EUR: '\u20AC',
  PLN: 'z\u0142',
  GBP: '\u00A3',
  RUB: '\u0440\u0443\u0431.',
  BRL: 'R$',
  CHF: 'CHF',
  SEK: 'kr',
  USD: '$',
  CLP: '$',
  BGN: 'лв.',
  CZK: 'Kč',
  DKK: 'kr.',
  HUF: 'Ft',
  NOK: 'kr',
  RON: 'Lei',
  HRK: 'kn',
  COP: '$'
};

function prependSymbol(amount, symbol, addSpace) {
  return `${symbol}${addSpace ? '\xA0' : ''}${amount}`;
}

function appendSymbol(amount, symbol, addSpace) {
  return `${amount}${addSpace ? '\xA0' : ''}${symbol}`;
}

function addSymbol(amount, symbol, { addSpace = true, prepend = false } = {}) {
  const symbolAddingFn = prepend ? prependSymbol : appendSymbol;
  return symbolAddingFn(amount, symbol, addSpace);
}

export function getCurrencyFormat(currency, locale) {
  const symbol = compose(
    defaultTo(currency),
    get(currency)
  )(CURRENCY_SYMBOLS);

  const { decimal: decimalSep, thousand: thousandSep } = getNumberFormat(
    locale
  );

  const currencyFormats =
    get(currency.toUpperCase(), CURRENCY_FORMATS) ||
    get('EUR', CURRENCY_FORMATS);

  const {
    prependSymbol: prepend,
    fractionalPrecision: currencyPrecision,
    addSpace
  } = get(locale, currencyFormats) || currencyFormats.default;

  return {
    decimalSep,
    thousandSep,
    prepend,
    currencyPrecision,
    addSpace,
    symbol
  };
}

export function shouldPrependSymbol(currency, locale) {
  const { prepend } = getCurrencyFormat(currency, locale);
  return prepend;
}

export function formatCurrency(amount, currency, locale) {
  const currencySymbol = compose(
    defaultTo(currency),
    get(currency)
  )(CURRENCY_SYMBOLS);
  const currencyFormat = getCurrencyFormat(currency, locale);

  const absAmount = Math.abs(amount);
  const sign = amount < 0 ? '-' : '';

  const formattedAmount = toCurrencyNumberFormat(absAmount, currencyFormat);
  const currencyString = addSymbol(
    formattedAmount,
    currencySymbol,
    currencyFormat
  );

  return `${sign}${currencyString}`;
}

function toCurrencyNumberFormat(number, currencyFormat) {
  const isAcceptedNumber = isNumber(number) || isString(number);
  const isCurrencyFormat = Object.keys(currencyFormat).length !== 0;
  if (!isAcceptedNumber || !isCurrencyFormat) {
    return number;
  }
  const numberFormat = {
    decimalSep: currencyFormat.decimalSep,
    thousandSep: currencyFormat.thousandSep,
    precision: currencyFormat.currencyPrecision
  };
  return formatNumber(number, numberFormat);
}

export function formatAmountForLocale(number, currency, locale) {
  const currencyFormat = getCurrencyFormat(currency, locale);
  return toCurrencyNumberFormat(number, currencyFormat);
}

export function currencyRegex(currency, locale) {
  const currencyFormat = get(currency, CURRENCY_FORMATS);
  const currencyLocaleFormat = get(locale, currencyFormat);
  const { fractionalPrecision } =
    currencyLocaleFormat || currencyFormat.default;

  const numberFormat = get(locale, NUMBER_SEPARATORS);
  const { decimal, thousand } = numberFormat;

  return currencyToRegex([thousand], fractionalPrecision, [decimal]);
}
