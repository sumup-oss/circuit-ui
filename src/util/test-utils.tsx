/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FunctionComponent } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, wait, act } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import ComponentsContext, {
  defaultComponents,
} from '../components/ComponentsContext';

export type RenderFn = (component: React.ReactElement, ...rest: any) => any;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const renderWithProviders = (renderer: RenderFn): RenderFn => (
  component,
  ...rest
): RenderFn => renderer(<WithProviders>{component}</WithProviders>, rest);

const render: RenderFn = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn = renderWithProviders(renderToStaticMarkup);
const create: RenderFn = (...args) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export {
  create,
  render,
  renderToHtml,
  renderHook,
  act,
  actHook,
  userEvent,
  wait,
  axe,
};
