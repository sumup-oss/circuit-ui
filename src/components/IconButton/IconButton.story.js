/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import IconButton from './IconButton';

export default {
  title: 'Components|Button/IconButton',
  component: IconButton,
  parameters: {
    jest: ['IconButton']
  }
};

const Icon = () => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="24"
    height="24"
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
    strokeWidth="2"
    strokeLinecap="round"
    strokeLinejoin="round"
    className="feather feather-eye"
  >
    <path d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z" />
    <circle cx="12" cy="12" r="3" />
  </svg>
);

export const base = () => (
  <IconButton label="Eye">
    <Icon />
  </IconButton>
);
