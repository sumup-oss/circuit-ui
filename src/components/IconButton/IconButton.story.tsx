/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { select, boolean, text } from '@storybook/addon-knobs';
import { Download } from '@sumup/icons';

import docs from '../Button/Button.docs.mdx';
import { IconButton } from './IconButton';

export default {
  title: 'Components/Button/IconButton',
  component: IconButton,
  parameters: {
    docs: { page: docs },
    jest: ['IconButton']
  }
};

export const Base = () => (
  <IconButton
    label={text('Label', 'Download')}
    variant={select('Variant', ['primary', 'secondary', 'tertiary'], 'primary')}
    size={select('Size', ['kilo', 'mega', 'giga'], 'mega')}
    stretch={boolean('Stretched', false)}
    disabled={boolean('Disabled', false)}
  >
    <Download />
  </IconButton>
);
