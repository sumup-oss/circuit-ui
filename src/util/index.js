import * as id from './id';
import * as numbers from './numbers';
import * as currency from './currency';
import * as sharedPropTypes from './shared-prop-types';
import * as sizes from './sizes';
import * as typeCheck from './type-check';

export { id, numbers, currency, sharedPropTypes, sizes, typeCheck };
