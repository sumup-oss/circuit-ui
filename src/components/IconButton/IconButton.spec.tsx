/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import { create, renderToHtml, axe } from '../../util/test-utils';

import { IconButton } from './IconButton';

describe('IconButton', () => {
  /**
   * Style tests.
   */
  it('should render with the default styles', () => {
    const actual = create(
      <IconButton label="A fake SVG">
        <div>SVG here</div>
      </IconButton>
    );
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <IconButton label="A fake SVG">
        <div>SVG here</div>
      </IconButton>
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
