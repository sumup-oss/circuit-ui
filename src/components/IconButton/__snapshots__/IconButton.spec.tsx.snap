// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should render with the default styles 1`] = `
.circuit-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-radius: 5px;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
}

.circuit-1:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-1:focus::-moz-focus-inner {
  border: 0;
}

.circuit-1:disabled,
.circuit-1[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-1:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-1:active {
  background-color: #EEF0F2;
  border-color: #5C656F;
}

.circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
  class="circuit-1"
  title="A fake SVG"
>
  <div
    role="presentation"
  >
    SVG here
  </div>
  <span
    class="circuit-0"
  >
    A fake SVG
  </span>
</button>
`;
